src/scheduler/pbs_sched.cpp

```diff
@@ -530,204 +506,139 @@ close_server(sched_svrconn *sconn)
-		pbs_disconnect(primary_sock);
-		pbs_disconnect(secondary_sock);
-		die(-1);
-	}
 
-	svr = static_cast<sched_svrconn *>(malloc(sizeof(sched_svrconn)));
-	if (!svr) {
-		log_err(errno, __func__, MEM_ERR_MSG);
-		die(-1);
-	}
+		svr_conns_primary =  static_cast<svr_conn_t *>(get_conn_svr_instances(clust_primary_sock));
+		svr_conns_secondary =  static_cast<svr_conn_t *>(get_conn_svr_instances(clust_secondary_sock));
 
-	svr->svrhost = strdup(svrhost);
-	if (!svr->svrhost) {
-		free(svr);
-		log_err(errno, __func__, MEM_ERR_MSG);
-		die(-1);
+		for (i = 0; i < num_conf_svrs; i++) {
+			if (svr_conns_primary[i].state == SVR_CONN_STATE_DOWN || svr_conns_primary[i].state == SVR_CONN_STATE_DOWN) {	
+				if (svr_conns_primary[i].state == SVR_CONN_STATE_DOWN)
+					clust_primary_sock = -1;
+				if (svr_conns_primary[i].state == SVR_CONN_STATE_DOWN)
+					clust_secondary_sock = -1;
+				break;
+			}
+		}
+
+		if (i != num_conf_svrs) {
+			/* If we reached here means one of the servers is down or not connected
+			 * we should go to the top of the loop again and call pbs_connect
+			 * Also wait for 2s for not to burn too much CPU
+			 */
+			sleep(2);
+			continue;
+		}
+
+		if (pbs_register_sched(sc_name) == 0) {
+			log_errf(pbs_errno, __func__, "Couldn't register the scheduler %s with the configured servers", sc_name);
+			/* wait for 2s for not to burn too much CPU, and then retry connection */
+			sleep(2);			
+			continue;
+		} else
+			registered = 1;
+		
 	}
-	svr->primary_sock = primary_sock;
-	svr->secondary_sock = secondary_sock;
-	svrs_newarr = add_ptr_to_array(servers, svr);
-	if (svrs_newarr == NULL) {
-		close_server(svr);
-		die(-1);
+
+
+	for (i = 0; i < get_num_servers(); i++) {
+		if (tpp_em_add_fd(poll_context, svr_conns_secondary[i].sd, EM_IN | EM_HUP | EM_ERR) < 0) {
+			log_errf(errno, __func__, "Couldn't add secondary connection to poll list for server %s", svr_conns_secondary[i].name);
+			die(-1);
+		}
 	}
-	servers = (sched_svrconn **)svrs_newarr;
-	log_eventf(PBSEVENT_ADMIN | PBSEVENT_FORCE, PBS_EVENTCLASS_SCHED, LOG_INFO, msg_daemonname, "Connected to server %s", svrhost);
 }
 
 /**
- * @brief Connect to all configured servers and populate servers list with
- *        connections to servers
+ * @brief Initialises event poll context for all the servers and also sched commands queue
  *
  * @return void
  */
 static void
-connect_servers(void)
+sched_svr_init(void)
 {
-	int max_connection = 1;
-	char *svrhost = pbs_default();
-
 	if (poll_context == NULL) {
-		poll_context = tpp_em_init(max_connection);
+		poll_context = tpp_em_init(get_num_servers());
 		if (poll_context == NULL) {
 			log_err(errno, __func__, "Failed to init cmd connections context");
 			die(-1);
 		}
+	}
 
-		sched_cmds = new_ds_queue();
-		if (sched_cmds == NULL) {
-			log_errf(errno, __func__, MEM_ERR_MSG);
-			die(-1);
-		}
+	qrun_list = static_cast<sched_cmd *>(malloc((get_num_servers() + 1) * sizeof(sched_cmd)));
```

Comment:
malloc returns a void pointer, does it still need to be casted?

Reply:
Same as the reason https://github.com/openpbs/openpbs/pull/2068#discussion_r510746731